core/include/utils/algorithm.hpp

```diff
@@ -18,35 +27,41 @@ namespace traccc {
  * to be able to transform the input type into the output type in one way or
  * another.
  *
- * @param I The input type of the algorithm
+ * @param A The input types of the algorithm
  * @param O The output type of the algorithm
  */
-template <typename I, typename O>
-class algorithm {
+template <typename R, typename... A>
+class algorithm<R(A...)> {
     public:
-    using input_type = I;
-    using output_type = O;
-
-    /**
-     * @brief Execute the algorithm, returning by value.
-     *
-     * Turn the object into a callable, accepting only an input and
-     * returning the output through value return.
-     *
-     * @param[in] i The input value to the algorithm
-     * @return An instance of the specified output type
-     */
-    virtual output_type operator()(const input_type& i) const = 0;
-
-    /**
-     * @brief Execute the algorithm, returning by input reference.
-     *
-     * Turn the object into a callable, accepting an input as well as an
-     * output parameter to write to.
-     *
-     * @param[in] i The input value to the algorithm
-     * @param[out] o The output of the algorithm
-     */
-    virtual void operator()(const input_type& i, output_type& o) const = 0;
+    using output_type = R;
+
+    using function_type = R(A...);
+
+    static_assert(
+        std::conjunction<rvalue_or_const_lvalue<A>...>::value,
+        "All arguments must be either affine types (rvalue references), or "
+        "immutable constant types (const lvalue references).");
+
+    virtual output_type operator()(A... args) const = 0;
 };
+
+template <typename A, typename B, typename C, typename... R>
+auto compose(std::function<std::remove_reference_t<B>(A)> f,
+             std::function<C(B)> g, R... rs) {
+    if constexpr (sizeof...(R) > 0) {
+        auto h = compose(g, rs...);
+        return [f, h](A&& i) { return h(f(std::forward<A>(i))); };
+    } else {
+        return [f, g](A&& i) { return g(f(std::forward<A>(i))); };
+    }
+}
+
+template <typename A, typename B, typename C>
+std::function<B(A&&)> side_effect(std::function<B(A)> f,
+                                  std::function<void(const C&)> s) {
+    return [=](A&& i) -> B {
+        s(i);
+        return f(std::forward<A>(i));
+    };
+}
```

Comment:
The arguments to `compose` can (sadly) have side-effects, but this function is specifically defined to add explicit side effects (in the forms of functions of type _A_ → void) to a pure algorithm. So, for example, if I want to print some string representing the type _A_, I could write a side-effecting function _f_ : _A_ → void, which I could not compose normally. Because for any function _g_ : _A_ → _B_, _g_ ∘ _f_ does not compose, and for another function _h_ : _A_ → _A_, _f_ ∘ _h_ : _A_ → void, which means that we cannot continue our computation meaningfully, because all we have is the a value of the unit type.

So what this function allows you to do is take a function _g_ : _A_ → _B_ and a function _f_ : _A_ → void, and essentially glue them into a new function _g'_ : _A_ → _B_ which incorporates the side-effects defined by _f_ into _g_ while still having the same type as _g_ originally had.

If you wanted to define the same behaviour using composition alone, you would need to define our side effecting function _f_ as _A_ → _A_, which works but is not as ergonomic (in my opinion) as the side-effect mechanism. In reverse this is much easier, because you can construct a function of exactly that type by gluing your side-effects to the identity function, which gives you exactly that type. But doing it this way is much easier that doing it the other way around! :smile:

Is it used anywhere except the tests, no, not yet! But my plan was to take some of our code like the measurement statistics and life them into a void-returning algorithm, so you could re-use that code in many different places very easily.

Reply:
Ok, sure if this is what you want. I'll out myself as a simpleton here, and state that I don't see the benefit, really, but let's go.

Reply:
Without a shadow of a doubt you are correct that I am over-complicating things. :see_no_evil:

Reply:
I'm with Paul here. This code seems to be a bit more complicated than what we would actually need in this project in practice. :stuck_out_tongue: But in the end it's just a couple of lines of helper functions, so I don't object to them either.